### PR TITLE
New resource 'azuredevops_serviceendpoint_permissions'

### PR DESCRIPTION
**Pull request by tiwood**
## All Submissions:

* [X] Have you added an explanation of what your changes do and why you'd like us to include them?
* [X] I have updated the documentation accordingly.
* [X] I have added tests to cover my changes.
* [x] All new and existing tests passed.
* [X] My code follows the code style of this project.
* [x] I ran lint checks locally prior to submission.
* [X] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?

This adds a new resource `azuredevops_serviceendpoint_permissions`

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [ ] Yes
- [X] No

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

## Any relevant logs, error output, etc?

N/A

## Other information

N/A

### CODE REVIEW

**Comment by tiwood** [comment]
@xuzhang3 not sure why tests are failing, it works on my end.

**Comment by xuzhang3** [comment]
Hi  @tiwood  You need change the permission name. The permission names returned from service cannot be mapped to  the locale configuration, string is case sensitive.   Change `CREATE` to `Create` etc.

**Comment by tiwood** [comment]
@xuzhang3, can you please check again? The tests seem to fail for a reason outside of the PR scope.

**Comment by xuzhang3** [comment]
Hi @tiwood  Can you share more service endpoint permission assignment? I try configure service endpoint permissions with UI and Terraform, the service endpoint permission looks like based on role assign [secure-a-service-connection](https://docs.microsoft.com/en-us/azure/devops/pipelines/library/service-endpoints?view=azure-devops&tabs=yaml#secure-a-service-connection). 
image


I'm wondering whether we can change the permission management to role assignment, but I cannot find any document about this.

**Comment by tiwood** [comment]
@xuzhang3, it seems this is a front-end feature/change.

We're currently using a custom built provider with the `azuredevops_serviceendpoint_permissions` resource and the following `HCL` results in these roles:

```hcl
#---------------------------------------------------------------------------
# This resource denies modification of the ARM Service Endpoints
resource "azuredevops_serviceendpoint_permissions" "project_power_users_deny_workload_arm_connection" {
  for_each            = local.connected_workloads_subscriptions_flat
  depends_on          = [azuredevops_serviceendpoint_azurerm.workload_arm_connection]
  project_id          = azuredevops_project.azdo_project.id
  service_endpoint_id = azuredevops_serviceendpoint_azurerm.workload_arm_connection[each.key].id
  principal           = azuredevops_group.project_power_users.descriptor
  permissions = {
    Use               = "allow"
    Administer        = "deny"
    Create            = "deny"
    ViewAuthorization = "allow"
    ViewEndpoint      = "allow"
  }
}
```

image

So basically if not `Administer == Allow` its a `User` otherwise `Administrator`. But that is just a guess.
If you ask me, we should stick with the way the API works not the front-end.

**Comment by xuzhang3** [comment]
@tiwood  It's weird that we grant the permissions but resulted in role assignment.  I need check this with the service team.

**Comment by tiwood** [comment]
@xuzhang3, just to clarify, a few weeks ago you would actually see the permissions. This role view has to be pretty new.

**Comment by avishnyakov** [comment]
Team, this is great work. What is left to move this work forward?

**Comment by tiwood** [comment]
@xuzhang3, any updates on this?
We're using this resource for months now without issue, but we would love to let go of our custom built version..

**Comment by xuzhang3** [comment]
@tiwood  I didn't get any response from service team. I think it is acceptable to merge this PR.  Plan to released with v0.1.5

**Comment by tmeckel** [comment]
@xuzhang3 will this PR be merged in the near future?

**Comment by xuzhang3** [comment]
Hi @tiwood  Can you help resolve  conflicts?

**Comment by tmeckel** [comment]
@xuzhang3 On behalf of @tiwood I rebased the pull request to the current main branch. 

In addition I added a `stateConf.WaitForState()` to `azuredevops/internal/service/permissions/utils/securityHelper.go` which will ensure that the (newly) set (updated) permissions are persisted in Azure DevOps **before** the function `SetPrincipalPermissions` returns to the caller. 

Although this will influence all code which handles permissions inside the provider and will prolong the runtime of every call to `SetPrincipalPermissions` it will make the provider much more resilient to race conditions while changing permissions.

@tiwood and I think that the pull request is ready for merge.

**Comment by tmeckel** [comment]
@xuzhang3  can we get this PR into the upcoming 0.1.8 release?

File: website/docs/r/serviceendpoint_permissions.html.markdown [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+  docker_username        = "sample"
+  docker_email           = "email@example.com"
+  docker_password        = "12345"
+  registry_type          = "DockerHub"
+}
+
+resource "azuredevops_serviceendpoint_permissions" "serviceendpoint-permissions" {
+  project_id         = azuredevops_project.project.id
+  principal          = azuredevops_group.project-readers.id
```
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_permissions.html.markdown]:
```suggestion
  principal          = data.azuredevops_group.project-readers.id
```

File: website/docs/r/serviceendpoint_permissions.html.markdown [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+data "azuredevops_group" "project-readers" {
+  project_id = azuredevops_project.project.id
+  name       = "Readers"
+}
+
+resource "azuredevops_serviceendpoint_permissions" "root-permissions" {
+  project_id  = azuredevops_project.project.id
+  principal   = azuredevops_group.project-readers.id
```
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_permissions.html.markdown]:
```suggestion
  principal   = data.azuredevops_group.project-readers.id
```

File: website/docs/r/serviceendpoint_permissions.html.markdown [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  docker_password        = "12345"
+  registry_type          = "DockerHub"
+}
+
+resource "azuredevops_serviceendpoint_permissions" "serviceendpoint-permissions" {
+  project_id         = azuredevops_project.project.id
+  principal          = azuredevops_group.project-readers.id
+  serviceendpoint_id = azuredevops_serviceendpoint_dockerregistry.dockerhubregistry.id
+  permissions = {
+    USE               = "allow"
```
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_permissions.html.markdown]: All the permission here is in upper case while the key name in [actionMap](https://github.com/microsoft/terraform-provider-azuredevops/blob/main/azuredevops/internal/service/permissions/utils/namespaces.go#L379) is in camel case, this seems a bug here.
Test script:
```
data "azuredevops_project" "project" {
  name               = "Sample Project"
}

data "azuredevops_group" "project-readers" {
  project_id = data.azuredevops_project.project.id
  name       = "Readers"
}

resource "azuredevops_serviceendpoint_permissions" "root-permissions" {
  project_id  = data.azuredevops_project.project.id
  principal   = data.azuredevops_group.project-readers.id
  permissions = {
    Use               = "allow"
    Administer        = "allow"
    Create            = "allow"
    ViewAuthorization = "allow"
    ViewEndpoint      = "allow"
  }
}

resource "azuredevops_serviceendpoint_dockerregistry" "dockerhubregistry" {
  project_id             = data.azuredevops_project.project.id
  service_endpoint_name  = "Sample Docker Hub"

  docker_username        = "sample"
  docker_email           = "email@example.com"
  docker_password        = "12345"
  registry_type          = "DockerHub"
}

resource "azuredevops_serviceendpoint_permissions" "serviceendpoint-permissions" {
  project_id         = data.azuredevops_project.project.id
  principal          = data.azuredevops_group.project-readers.id
  serviceendpoint_id = azuredevops_serviceendpoint_dockerregistry.dockerhubregistry.id
  permissions = {
    Use               = "allow"
    Administer        = "deny"
    Create            = "deny"
    ViewAuthorization = "allow"
    ViewEndpoint      = "allow"
  }
}
```


Error:
```
Error: Invalid permission [VIEWENDPOINT]

  on main.tf line 25, in resource "azuredevops_serviceendpoint_permissions" "root-permissions":
  25: resource "azuredevops_serviceendpoint_permissions" "root-permissions" {



Error: Invalid permission [Viewauthorization]

  on main.tf line 47, in resource "azuredevops_serviceendpoint_permissions" "serviceendpoint-permissions":
  47: resource "azuredevops_serviceendpoint_permissions" "serviceendpoint-permissions" {


```

**Review by xuzhang3 — CHANGES_REQUESTED** [review]
@tmeckel  A few changes request, others looks good to me.

File: website/docs/r/serviceendpoint_permissions.html.markdown (hunk 3 above)
Comment (tmeckel) [reply website/docs/r/serviceendpoint_permissions.html.markdown]: @xuzhang3 will have a look during the day

Comment (tmeckel) [reply website/docs/r/serviceendpoint_permissions.html.markdown]: @xuzhang3 checked the security namespace. You're correct, the capitalization for the permissions inside the documentation is incorrect

image

Comment (tmeckel) [reply website/docs/r/serviceendpoint_permissions.html.markdown]: @xuzhang3 fixed

**Comment by xuzhang3** [comment]
```log
==> Checking that code complies with gofmt requirements...
==> Sourcing .env file if avaliable
if [ -f .env ]; then set -o allexport; . ./.env; set +o allexport; fi; \
        TF_ACC=1 go test -count=1 -tags "all" $(go list ./azuredevops/internal/acceptancetests |grep -v 'vendor') -v -run=TestAccServiceEndpointPermissions -timeout 120m
=== RUN   TestAccServiceEndpointPermissions_SetPermissions
--- PASS: TestAccServiceEndpointPermissions_SetPermissions (132.97s)
=== RUN   TestAccServiceEndpointPermissions_UpdatePermissions
--- PASS: TestAccServiceEndpointPermissions_UpdatePermissions (153.12s)
PASS
ok      github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/acceptancetests        287.243s

```

**Comment by xuzhang3** [comment]
LGTM

**Comment by aksharpanchal** [comment]
How to handle permissions at organizational level? this handles adding permissions to Security -> User permissions -> Project level. Is adding permissions at organizational level handled by a different resource type? 

image

**Comment by xuzhang3** [comment]
@aksharpanchal this is not supported yet.

**Pull request merged**